File: core/data/src/main/java/me/nalab/core/data/survey/ChoiceFormQuestion.java
package me.nalab.core.data.survey;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.OneToMany;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

@Entity
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
public class ChoiceFormQuestion extends FormQuestion {

	@OneToMany(mappedBy = "choiceFormQuestion", fetch = FetchType.LAZY)
	private List<Choice> choiceList;

	@Column(name = "max_selection_count")
	private Integer maxSelectionCount;

	@Enumerated(EnumType.STRING)
	@Column(name = "choice_question_type")
	private ChoiceFormQuestionType choiceFormQuestionType;

}

File: src/main/kotlin/domain/survey/ChoiceFormQuestion.kt
package domain.survey

import me.nalab.survey.domain.survey.exception.DuplicatedOrderException
import javax.persistence.*

@Entity
class ChoiceFormQuestion(
    id: Long,
    title: String,
    order: Int,
    questionType: QuestionType,
    survey: Survey,

    @OneToMany(
        mappedBy = "choiceFormQuestion",
        fetch = FetchType.LAZY,
        cascade = [CascadeType.PERSIST, CascadeType.MERGE],
    )
    val choices: MutableList<Choice>,

    @Column(name = "max_selection_count")
    val maxSelectableCount: Int,

    @Enumerated(EnumType.STRING)
    @Column(name = "choice_question_type")
    val choiceFormQuestionType: ChoiceFormQuestionType,
) : FormQuestionable(id, title, order, questionType, survey) {

    init {
        choices.sorted()
        validNoDuplicatedChoiceOrder(choices)
    }

    private fun validNoDuplicatedChoiceOrder(choiceList: List<Choice>) {
        val orders = HashSet<Int>()
        choiceList.forEach { choice ->
            val order: Int = choice.order
            if (orders.contains(order)) {
                throw DuplicatedOrderException(order, orders)
            }
            orders.add(order)
        }
    }
}
